lib/spack/llnl/util/link_tree.py

```diff
@@ -150,15 +151,27 @@ def visit_file(self, root: str, rel_path: str, depth: int) -> None:
                 )
             )
         elif proj_rel_path in self.files:
-            # In some cases we can resolve file-file conflicts
+            # Resolve file-file conflicts when two files are identical: this deals with cases of
+            # of symlink to symlink, symlink to regular file and hardlinked files, where in
+            # principle the file we pick is arbitrary. However, for environment views that copy,
+            # we want the file to be copied, not the symlink, so we pick the underlying file, which
+            # should be the current, assuming proper ordering.
+            # NOTE: in principle filecmp could be used too as a more generic solution, but that is
+            # relatively slow, and so far this visitor has only used file metadata, not contents.
             src_a_root, src_a_relpath = self.files[proj_rel_path]
-            self.file_conflicts.append(
-                MergeConflict(
-                    dst=proj_rel_path,
-                    src_a=os.path.join(src_a_root, src_a_relpath),
-                    src_b=os.path.join(root, rel_path),
+
+            src_a = os.path.join(src_a_root, src_a_relpath)
+            src_b = os.path.join(root, rel_path)
+
+            if os.path.samefile(src_a, src_b):
+                # Delete and re-insert to retain correct self.files ordering.
+                del self.files[proj_rel_path]
+                self.files[proj_rel_path] = (root, rel_path)
```

Comment:
The tests cover it.

Reply:
>> Other option: pick whichever one is a file, and use that

> First, that doesn't work when both are links. 

If they're both links, it should be fine to use either

> Second, realpath triggers O(depth + chain length) readlink calls

Could we `lstat` the files to determine whether they are links (which I think would involve fewer calls than `realpath`).

```
if samefile() and xor(islink(x) for x in src_a, src_b):
  # one of these is a symlink to the other, use the original file
```

As a general question: Doesn't samefile have to do the same thing as realpath? How could `samefile` work if it's not chasing down the link targets (transitively)?

Reply:
The visitor already has `visit_symlink`.

> If they're both links, it should be fine to use either

Then why suggest to distinguish instead of leaving it at `os.samefile`?

> Doesn't samefile have to do the same thing as realpath? How could samefile work if it's not chasing down the link targets (transitively)?

The kernel can avoid chasing through file system caches you can't have in userland, cause there you don't know how to invalidate them. But also it's just single syscall versus many sycalls.

Reply:
> The visitor already has visit_symlink.

In that case the check would be:

```
if samefile() and islink(src_a):
  # src_a is a symlink to src_b, which is a file: use the file.
  # We do this because src_a would become self-referential
  # if relocated relative to the link_tree's root
```

I think this comment would be useful compared to

```
            # Resolve file-file conflicts when two files are identical: this deals with cases of
            # of symlink to symlink, symlink to regular file and hardlinked files, where in
            # principle the file we pick is arbitrary. However, for environment views that copy,
            # we want the file to be copied, not the symlink, so we pick the underlying file, which
            # should be the current, assuming proper ordering.
```

(the former gets into the why, and avoids caring about the cases we don't need to change)

> Then why suggest to distinguish instead of leaving it at os.samefile

Doing something that doesn't need to be done (i.e. replacing with `src_b` when we don't need to) is confusing in this case: it obfuscates the reasons for doing it.

I'm open to other approaches to making this less confusing (e.g. the comment I suggested above would help).
Perhaps (also) putting this in the docstring for `SourceMergeVisitor` (and `LinkTree.merge`) would help: `test_source_merge_visitor_does_not_register_identical_file_conflicts` enforces that behavior for the tests, but the behavior of `merge` for these types of issues seems like something that ought to be explicit in the method's description.

Reply:
To be honest I don't think `islink(src_a)` improves things _that_ much, cause you still need to explain the assumption on order: dependents before dependencies, symlinks only go from dependent -> dependency. 

The conflicting symlink to conflicting symlink to conflicting file case is just _so_ incredibly niche, I don't get why you care about "not doing something that doesn't need to be done". Also note that none of this is strictly necessary for copy type environments.

I'll try and write a concise and less technical comment based on your suggestion, that hopefully resolves the confusion.

Reply:
> To be honest I don't think islink(src_a) improves things that much,

This is convoluted and I'm trying to find any way I can to make it less so. If you think this makes it even the slightest bit clearer, then it's worthwhile. Without checking `islink` the comment I suggested is imprecise and therefore less useful.

Zooming out, doing this in `add_files_to_view` would have been more clear because the concerns would have been addressed precisely where they arise. Another mechanism would be for example registering a conflict-resolution callback with the visitor that views could add their own implementation to. There's plenty of angles that avoid the requirement for discussing actions that only make sense for isolated users of this, so we could also consider those if it's impossible to make this particular approach more clear.

> cause you still need to explain the assumption on order: dependents before dependencies, symlinks only go from dependent -> dependency.

I thought the `islink(src_a)` part of `if samefile and islink(src_a)` would make it work in both directions (we only replace with `src_b` if `src_b` is the file.

Reply:
> would make it work in both directions 

True. I've done the reverse now: if we currently have a regular file, use that. That would be a false positive for hardlinks, but frankly cannot be bothered with that, cause it really doesn't matter what file is linked.